### PR TITLE
httputil,store: extract retry code to httputil, reorg usages

### PR DESCRIPTION
**Pull request by pedronis**
This extracts the retry code in store to httputil so that it can reused outside of store, it reorgs and generalises some of its details.

### CODE REVIEW

File: store/auth.go [hunk 1]
```diff
@@ -341,8 +295,8 @@ func requestDeviceSession(serialAssertion, sessionRequest, previousSession strin
 		headers["X-Device-Authorization"] = fmt.Sprintf(`Macaroon root="%s"`, previousSession)
 	}
 
-	_, err = retryPostRequest(MyAppsDeviceSessionAPI, headers, deviceJSONData, func(ok bool, resp *http.Response) error {
-		if ok {
+	_, err = retryPostRequest(MyAppsDeviceSessionAPI, headers, deviceJSONData, func(resp *http.Response) error {
+		if resp.StatusCode == 200 || resp.StatusCode == 202 {
```
Comment (stolowski) [inline store/auth.go]: It would make sense to use http.StatusOK and http.StatusCreated constants here for consistency.

File: store/store.go [hunk 2]
```diff
@@ -1606,9 +1580,12 @@ func (s *Store) Assertion(assertType *asserts.AssertionType, primaryKey []string
...
 
 	var asrt asserts.Assertion
-	resp, err := s.retryRequest(context.TODO(), s.client, reqOptions, user, func(ok bool, resp *http.Response) error {
+
+	resp, err := httputil.RetryRequest(reqOptions.URL.String(), func() (*http.Response, error) {
+		return s.doRequest(context.TODO(), s.client, reqOptions, user)
+	}, func(resp *http.Response) error {
 		var e error
-		if ok {
+		if resp.StatusCode == 200 {
```
Comment (stolowski) [inline store/store.go]: Use http.StatusOK.

File: store/userinfo.go [hunk 3]
```diff
@@ -51,47 +49,38 @@ type User struct {
...
-		switch resp.StatusCode {
-		case 200: // good
-		case 404:
-			return nil, fmt.Errorf("cannot find user %q", email)
-		default:
-			return nil, respToError(resp, fmt.Sprintf("look up user %q", email))
+	resp, err := httputil.RetryRequest(ssourl, func() (*http.Response, error) {
+		return httpClient.Get(ssourl)
+	}, func(resp *http.Response) error {
+		if resp.StatusCode != 200 {
```
Comment (stolowski) [inline store/userinfo.go]: and here.

File: store/userinfo.go [hunk 4]
```diff
@@ -51,47 +49,38 @@ type User struct {
...
+	}
 
-		return &User{
-			Username:         v.Username,
-			SSHKeys:          v.SSHKeys,
-			OpenIDIdentifier: v.OpenIDIdentifier,
-		}, nil
+	switch resp.StatusCode {
+	case 200: // good
+	case 404:
```
Comment (stolowski) [inline store/userinfo.go]: and here too.

Comment (stolowski) [inline store/userinfo.go]: Also, I agree with @chipaca about dropping Context.. Not sure that's a good idea, perhaps it could be moved to requestOptionis?

Comment (pedronis) [inline store/userinfo.go]: as I said I find it a bit immaterial either way at this point, but I'm ok putting it back

File: store/auth.go (hunk 1 above)
Comment (pedronis) [inline store/auth.go]: we are actually not that consistent, I think it merits its own PR to fix that

File: store/store.go (hunk 2 above)
Comment (pedronis) [inline store/store.go]: same point

File: store/userinfo.go (hunk 3 above)
Comment (pedronis) [inline store/userinfo.go]: same point

File: store/userinfo.go (hunk 4 above)
Comment (pedronis) [inline store/userinfo.go]: context is back

File: store/auth.go (hunk 1 above)
Comment (pedronis) [inline store/auth.go]: after team discussion we agreed to do the other way, using numerical codes

**Pull request merged**